pkgs/build-support/build-fhsenv-bubblewrap/rootfs-builder/src/main.rs

```diff
@@ -0,0 +1,274 @@
+    };
+
+    let Ok(Object::Elf(e)) = goblin::Object::parse(&buf) else {
+        return false;
+    };
+
+    !e.is_64
+}
+
+fn remap_multilib_path(root: &Path, p: &Path) -> Option<PathBuf> {
+    if let Ok(no_lib) = p.strip_prefix("lib/") {
+        let full = root.join(p);
+
+        let libdir = if is_32_bit(&full) { "lib32" } else { "lib64" };
+
+        return Some(PathBuf::from("usr/").join(libdir).join(no_lib));
+    }
+
+    if p.starts_with("etc/") || p.starts_with("opt/") {
+        return Some(p.into());
+    }
+
+    if p.starts_with("share/") || p.starts_with("include/") {
+        return Some(PathBuf::from("usr/").join(p));
+    }
+
+    None
+}
+
+fn build_plan(
+    paths: HashMap<PathBuf, PriorityKey>,
+    paths32: HashMap<PathBuf, PriorityKey>,
+) -> anyhow::Result<HashMap<PathBuf, PathBuf>> {
+    let candidates_native = collect_candidate_paths(paths, remap_native_path)?;
+    let candidates_32 = collect_candidate_paths(paths32, remap_multilib_path)?;
+
+    let mut all_candidates: HashMap<_, Vec<_>> = HashMap::new();
+
+    for map in [candidates_native, candidates_32] {
+        for (path, candidates) in map {
+            all_candidates
+                .entry(path)
+                .or_default()
+                .extend_from_slice(&candidates);
+        }
+    }
+
+    let mut final_plan: HashMap<PathBuf, PathBuf> = HashMap::new();
+    for (path, candidates) in all_candidates {
+        let best = candidates
+            .into_iter()
+            .min_by_key(|&CandidatePath { priority, .. }| priority)
+            .ok_or(anyhow!("candidate list empty"))?;
+        final_plan.insert(path, best.root.join(best.relative));
+    }
+
+    Ok(final_plan)
+}
+
+fn build_env(out: &Path, plan: HashMap<PathBuf, PathBuf>) -> anyhow::Result<()> {
+    fs::create_dir_all(out)?;
+
+    for (dest, src) in plan {
+        let full_dest = out.join(&dest);
+        let dest_dir = full_dest
+            .parent()
+            .ok_or(anyhow!("destination directory is root"))
+            .with_context(|| {
+                format!("When trying to determine destination directory for {full_dest:?}")
+            })?;
+        fs::create_dir_all(dest_dir)
+            .with_context(|| format!("When trying to create directory {dest_dir:?}"))?;
+        ufs::symlink(&src, &full_dest)
+            .with_context(|| format!("When symlinking {src:?} to {full_dest:?}"))?;
+    }
+
+    Ok(())
+}
+
+fn main() -> anyhow::Result<()> {
+    let filename = env::var("NIX_ATTRS_JSON_FILE")?;
+
+    let reader = File::open(filename)?;
+    let buf_reader = BufReader::new(reader);
+
+    let attrs: StructuredAttrsRoot = serde_json::from_reader(buf_reader)?;
+
+    let refs = build_reference_map(attrs.graph);
+
+    let closure = collect_paths_with_priority(&attrs.paths, 1, &refs)?;
+    let closure32 = collect_paths_with_priority(&attrs.paths32, 2, &refs)?;
+
+    let plan = build_plan(closure, closure32)?;
+
+    let out_dir = env::var("out")?;
```

Comment:
Consider adding context to the `?` failures here.

Reply:
Those are the "this should never happen" types of failures, those environment variables are set by Nix/stdenv itself.